community/kernel/src/main/java/org/neo4j/kernel/impl/store/kvstore/AbstractKeyValueStore.java

```diff
@@ -300,14 +300,27 @@ private long rotate( boolean force ) throws IOException
         {
             try( RotationState<Key> rotation = this.rotation )
             {
-                final long version = rotation.rotationVersion();
-                ProgressiveState<Key> next = rotation.rotate( force, rotationStrategy, rotationTimerFactory,
-                        value -> updateHeaders( value, version ) );
-                try ( LockWrapper ignored = writeLock( updateLock ) )
+                try
                 {
-                    state = next;
+                    final long version = rotation.rotationVersion();
+                    ProgressiveState<Key> next = rotation.rotate( force, rotationStrategy, rotationTimerFactory,
+                            value -> updateHeaders( value, version ) );
+                    try ( LockWrapper ignored = writeLock( updateLock ) )
+                    {
+                        state = next;
+                    }
+                    return version;
+                }
+                catch ( Throwable t )
+                {
+                    // Rotation failed. Here we assume that rotation state remembers this so that closing it
+                    // won't close the state as it was before rotation began, which we're reverting to right here.
```

Comment:
You said you like option `1`, but you have implemented option `2`.

There are test failures that look related, so I'm not so sure it is safe to use the same flag for two purposes.

Before your changes this flag was used to ensure that we could still close the store even if the final rotation failed.

In your changes you want to ensure that we don't close the previous state if non-final rotation fails.

These two are quite opposing objectives, so it seems like it would be a bad idea to use the same flag (`RotationState.Rotation.failed`) for both purposes.